src/host.cpp

```diff
@@ -847,113 +818,42 @@ namespace rhost {
-#ifdef TRACE_JSON
-            logf("==> %s\n\n", json.c_str());
-#endif
-
-            picojson::value value;
-            auto err = picojson::parse(value, json);
-            if (!err.empty()) {
-                fatal_error("Malformed incoming message: %s", err.c_str());
-            }
-
-            if (value.is<picojson::null>()) {
-                terminate("Shutdown request received.");
-            }
-
-            if (!value.is<picojson::array>()) {
-                fatal_error("Message must be an array.");
-            }
-
-            auto& array = value.get<picojson::array>();
-            if (array.size() < 1) {
-                fatal_error("Message must be of the form [[id, name, blob_count], ...].");
-            }
-
-            auto& header = array[0].get<picojson::array>();
-            if (header.size() < 3 || !header[0].is<std::string>() || !header[1].is<std::string>() || !header[2].is<double>()) {
-                fatal_error("Message must be of the form [[id, name, blob_count], ...].");
-            }
-
-            message incoming = {};
-            incoming.id = header[0].get<std::string>();
-            incoming.name = header[1].get<std::string>();
-            if (incoming.name.empty()) {
-                fatal_error("Message must have a non-empty name.");
-            }
+        void ws_message_handler(websocketpp::connection_hdl hdl, ws_connection_type::message_ptr msg) {
+            const auto& payload = msg->get_payload();
 
-            incoming.blob_count = static_cast<size_t>(header[2].get<double>());
-            if (incoming.blob_count < 0) {
-                fatal_error("blob_count must be >= 0.");
-            }
+            auto incoming = message::parse(payload);
+            log_message("==>", incoming.id(), incoming.request_id(), incoming.name(), incoming.json(), incoming.blob());
 
-            if (incoming.name[0] == ':') {
-                if (header.size() < 4 || !header[3].is<std::string>()) {
-                    fatal_error("Response message must be of the form [[id, name, blob_count, request_id], ...].");
+            if (incoming.is_response()) {
+                std::lock_guard<std::mutex> lock(response_mutex);
+                assert(response_state != RESPONSE_RECEIVED);
+                if (response_state == RESPONSE_UNEXPECTED) {
+                    fatal_error("Unexpected incoming client response.");
                 }
 
-                incoming.request_id = header[3].get<std::string>();
+                response = std::move(incoming);
+                response_state = RESPONSE_RECEIVED;
+                unblock_message_loop();
+                return;
             }
 
-            auto args = array.begin() + 1;
-            incoming.args.assign(args, array.end());
-
-            if (incoming.name == "!/") {
+            std::string name = incoming.name();
+            if (name == "!End") {
+                terminate("Shutdown request received.");
+            } else if (name == "!/") {
```

Comment:
Yeah. I want to do a big refactoring of this code, splitting it into transport (just handles generic messages and invokes handlers via map) and actual host code. I actually started this change that way, but it's too big for this milestone at this point.